index.js

```diff
@@ -1,174 +1,221 @@
+        opts.codec = codec
+      }
     } else {
-      result[key] = value
+      if (!opts.cid) throw new Error('Cannot create block instance without cid or codec')
+      opts.codec = Block.codecs.get(opts.cid.code)
     }
-  })
-
-  class Block {
-    constructor (opts) {
-      if (!opts) throw new Error('Block options are required')
-      if (opts.cid) opts.cid = CID.asCID(opts.cid)
-      if (typeof opts.codec === 'number') {
-        opts.code = opts.codec
-        opts.codec = multicodec.get(opts.code).name
-      }
-      if (typeof opts.source === 'undefined' &&
-          typeof opts.data === 'undefined') {
-        throw new Error('Block instances must be created with either an encode source or data')
-      }
-      if (typeof opts.source !== 'undefined' && !opts.codec && !opts.code) {
-        throw new Error('Block instances created from source objects must include desired codec')
-      }
-      if (opts.data && !opts.cid && !opts.codec && !opts.code) {
-        throw new Error('Block instances created from data must include cid or codec')
-      }
-      if (!opts.cid && !opts.algo) opts.algo = 'sha2-256'
-      // Do our best to avoid accidental mutations of the options object after instantiation
-      // Note: we can't actually freeze the object because we mutate it once per property later
-      opts = Object.assign({}, opts)
-      Object.defineProperty(this, 'opts', readonly(opts))
+    if (typeof opts.source === 'undefined' &&
+        typeof opts.data === 'undefined') {
+      throw new Error('Block instances must be created with either an encode source or data')
     }
-
-    source () {
-      if (this.opts.cid || this.opts.data ||
-          this._encoded || this._decoded) return null
-      if (!this.opts.source) return null
-      return this.opts.source
+    if (typeof opts.source !== 'undefined' && !opts.codec && !opts.code) {
+      throw new Error('Block instances created from source objects must include desired codec')
     }
-
-    async cid () {
-      if (this.opts.cid) return this.opts.cid
-      const hash = await multihash.hash(this.encodeUnsafe(), this.opts.algo)
-      const cid = CID.create(1, this.code, hash)
-      this.opts.cid = cid
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return cid
+    if (opts.data && !opts.cid && !opts.codec && !opts.code) {
+      throw new Error('Block instances created from data must include cid or codec')
     }
+    opts.hasher = opts.hasher || sha256
+    // Do our best to avoid accidental mutations of the options object after instantiation
+    // Note: we can't actually freeze the object because we mutate it once per property later
+    opts = Object.assign({}, opts)
+    Object.defineProperty(this, 'opts', readonly(opts))
+    Object.defineProperty(this, 'asBlock', readonly(this))
+  }
 
-    get codec () {
-      if (this.opts.code) {
-        this.opts.codec = multicodec.get(this.opts.code).name
+  get hasher () {
```

Comment:
I agree, there’s a bigger refactor that I’m going to want to do to the construct and class methods that i haven’t done yet. I wanted to first see the existing implementation ported over without much API changes, but I’m still debating with myself whether or not i’m going to tackle the bigger constructor refactor *before* i get this merged or just do it now so that I can combine the breaking changes to all the APIs.